File: Strivers-A2Z-DSA-Sheet/13. Graphs/5. MST Problems/python/07. Making Large Island.py
"""
Question:
Given an n x n binary matrix grid, change at most one 0 to 1.
Return the size of the largest island after the operation.

Approach:
- Use Disjoint Set to connect all 1-cells into islands.
- For each 0-cell, check the unique neighboring island sizes and compute potential island size.

Time Complexity: O(n^2)
Space Complexity: O(n^2)
"""


class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n

    def find_u_par(self, node):
        if self.parent[node] == node:
            return node
        self.parent[node] = self.find_u_par(self.parent[node])
        return self.parent[node]

    def union_by_size(self, u, v):
        uP, vP = self.find_u_par(u), self.find_u_par(v)
        if self.size[uP] > self.size[vP]:
            self.parent[vP] = uP
            self.size[uP] += self.size[vP]
        else:
            self.parent[uP] = vP
            self.size[vP] += self.size[uP]


def is_land(i, j, grid):
    return 0 <= i < len(grid) and 0 <= j < len(grid) and grid[i][j] == 1


def largest_island(grid):
    n = len(grid)
    dr = [1, -1, 0, 0]
    dc = [0, 0, 1, -1]
    djs = DisjointSet(n * n)

    for i in range(n):
        for j in range(n):
            if grid[i][j] == 1:
                cell_id = i * n + j
                for d in range(4):
                    ni, nj = i + dr[d], j + dc[d]
                    if is_land(ni, nj, grid):
                        nid = ni * n + nj
                        if djs.find_u_par(cell_id) != djs.find_u_par(nid):
                            djs.union_by_size(cell_id, nid)

    ans = float('-inf')
    for i in range(n):
        for j in range(n):
            if grid[i][j] == 0:
                st = set()
                for d in range(4):
                    ni, nj = i + dr[d], j + dc[d]
                    if is_land(ni, nj, grid):
                        st.add(djs.find_u_par(ni * n + nj))
                siz = sum(djs.size[u] for u in st)
                ans = max(ans, siz + 1)

    for cell_no in range(n * n):
        ans = max(ans, djs.size[djs.find_u_par(cell_no)])

    return ans

File: Strivers-A2Z-DSA-Sheet/13. Graphs/5. MST Problems/javascript/07. Making Large Island.js
/*
Question:
Given an n x n binary matrix grid, change at most one 0 to 1.
Return the size of the largest island after the operation.

Approach:
- Use Disjoint Set to connect all 1-cells into islands.
- For each 0-cell, check the unique neighboring island sizes and compute potential island size.

Time Complexity: O(n^2)
Space Complexity: O(n^2)
*/

class DisjointSet {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.size = new Array(n).fill(1);
    }

    findUPar(node) {
        if (this.parent[node] === node) return node;
        return this.parent[node] = this.findUPar(this.parent[node]);
    }

    unionBySize(u, v) {
        let uP = this.findUPar(u), vP = this.findUPar(v);
        if (this.size[uP] > this.size[vP]) {
            this.parent[vP] = uP;
            this.size[uP] += this.size[vP];
        } else {
            this.parent[uP] = vP;
            this.size[vP] += this.size[uP];
        }
    }
}

function isLand(i, j, grid) {
    return i >= 0 && i < grid.length && j >= 0 && j < grid.length && grid[i][j] === 1;
}

function largestIsland(grid) {
    let n = grid.length;
    let dr = [1, -1, 0, 0];
    let dc = [0, 0, 1, -1];
    let djs = new DisjointSet(n * n);

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === 1) {
                let id = i * n + j;
                for (let d = 0; d < 4; d++) {
                    let ni = i + dr[d], nj = j + dc[d];
                    if (isLand(ni, nj, grid)) {
                        let nid = ni * n + nj;
                        if (djs.findUPar(id) !== djs.findUPar(nid)) {
                            djs.unionBySize(id, nid);
                        }
                    }
                }
            }
        }
    }

    let ans = -Infinity;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === 0) {
                let st = new Set();
                for (let d = 0; d < 4; d++) {
                    let ni = i + dr[d], nj = j + dc[d];
                    if (isLand(ni, nj, grid)) {
                        st.add(djs.findUPar(ni * n + nj));
                    }
                }
                let siz = 0;
                for (let u of st) siz += djs.size[u];
                ans = Math.max(ans, siz + 1);
            }
        }
    }

    for (let cellNo = 0; cellNo < n * n; cellNo++) {
        ans = Math.max(ans, djs.size[djs.findUPar(cellNo)]);
    }
    return ans;
}

